## Machine Learning Record Mining

Project to create a pipeline that uses GeoDeepDive's output to find Unaquired Sites for Neotoma.

Using NLP parsed text and a Data Science approach, identify whether a paper is suitable for Neotoma and detect features such as 'Site Name', 'Location', 'Age Span' and 'Site Descriptions'.

In [1]:
# Loading libraries

import numpy as np
import pandas as pd
import csv
import psycopg2
import re

# Load Postgres Server
from src.config import config

In [161]:
# Options for DF display
#pd.set_option('display.max_colwidth', 50)
#pd.set_option('display.max_rows', 10)

pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)
pd.set_option('display.max_colwidth', -1)

## Loading and viewing the Data

### Loading NLP Sentences

In [3]:
# Connect to PostgreSQL server from terminal:
# pg_ctl -D PSQL_Data -l logfile start

try:
    params = config()
    # Connect to the PostgreSQL database
    conn = psycopg2.connect(**params)
    # Create a new cursor
    cur = conn.cursor()
    nlp_sentences = pd.read_sql_query('''SELECT * FROM sentences;''', conn)
    # Close the cursor and connection to so the server can allocate
    # bandwidth to other requests
    cur.close()
    conn.close()
    print('Data obtained from PostgreSQL')
    
# If no SQL db, load from a file
except:
    header_list = ["_gddid", "sentid", "wordidx", "words", "part_of_speech", "special_class", 
               "lemmas", "word_type", "word_modified"]
    nlp_sentences = pd.read_csv("../Do_not_commit_data/sentences_nlp352", sep='\t', names = header_list)
    nlp_sentences = nlp_sentences.replace('"', '', regex = True)\
                                 .replace('\{', '', regex = True)\
                                 .replace('}', '', regex = True)\
                                 .replace(',', ',', regex = True)
    nlp_sentences['wordidx']= nlp_sentences['wordidx'].str.split(",")
    nlp_sentences['words']= nlp_sentences['words'].str.split(",")
    nlp_sentences['poses']= nlp_sentences['poses'].str.split(",")
    nlp_sentences['ners']= nlp_sentences['ners'].str.split(",")
    nlp_sentences['lemmas']= nlp_sentences['lemmas'].str.split(",")
    nlp_sentences['dep_paths']= nlp_sentences['dep_paths'].str.split(",")
    nlp_sentences['dep_parents']= nlp_sentences['dep_parents'].str.split(",")
    print('Data obtained from text file')

Data obtained from PostgreSQL


In [4]:
nlp_sentences.head(3)

_gddid  sentid  \
0  54b43266e138239d8684efed       1   
1  54b43266e138239d8684efed       2   
2  54b43266e138239d8684efed       3   

                                             wordidx  \
0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   
1            [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]   
2  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...   

                                               words  \
0  [Available, online, at, www.sciencedirect.com,...   
1  [The, Chihuahueños, Bog, record, extends, to, ...   
2  [An, Artemisia, steppe, ,, then, an, open, Pic...   

                                      part_of_speech  \
0  [JJ, NN, IN, NNP, NNP, NNP, CD, -LRB-, CD, -RR...   
1  [DT, NNP, NN, NN, VBZ, TO, IN, CD, JJ, NN, NN, .]   
2  [DT, NNP, NN, ,, RB, DT, JJ, NNP, NN, VBD, IN,...   

                                       special_class  \
0  [O, O, O, O, O, O, NUMBER, O, DATE, O, NUMBER,...   
1   [O, O, O, O, O, O, O, NUMBER, O, DURATION, O, O]   
2  [O, O, O, O, O, O, O, LOCATION, LOCATION, O, O...   

                                              lemmas  \
0  [available, online, at, www.sciencedirect.com,...   
1  [the, Chihuahueños, bog, record, extend, to, o...   
2  [a, Artemisia, steppe, ,, then, a, open, Picea...   

                                           word_type  \
0  [dep, dep, dep, dep, dep, dep, dep, , dep, , d...   
1  [det, compound, compound, nsubj, , case, amod,...   
2  [det, compound, nsubj, , advmod, det, amod, co...   

                                       word_modified  
0  [218, 218, 218, 218, 218, 218, 218, 0, 218, 0,...  
1          [4, 4, 4, 5, 0, 11, 11, 11, 11, 11, 5, 0]  
2  [3, 3, 10, 0, 9, 9, 9, 9, 10, 0, 14, 14, 14, 1...

In [5]:
# Running this takes forever. Do not run unless necessary
# nlp_sentences.describe(include = 'all')

In [6]:
avg_sentences = nlp_sentences.groupby('_gddid')['_gddid','sentid'].transform('max')
avg_sentences.drop_duplicates(subset='_gddid', keep='first', inplace=True)
avg_sentences.describe()

sentid
count   218.000000
mean    513.600917
std     232.460891
min       2.000000
25%     397.250000
50%     504.000000
75%     638.250000
max    1330.000000

From the Summary Statistics, despite having 111,965 sentences, there are only 218 unique articles (`_gddid`). The article with most sentences, had 1330 sentences; that's the max number found in `sentid`.  
In order to know the average of sentences, I grouped by `_gddid` and then, obtained the number of sentences per article. After that, I just got the mean through a Summary of statistics to know that on average, each article has 514 sentences.

### Loading Bibliography Data

In [7]:
import json
with open('../Do_not_commit_data/bibjson', 'r') as f:
    bib_dict = json.load(f)

In [8]:
# Normalizing data so that we have access to the 'identifier'

# TODO Load into SQL server and connect through SQL
bibliography = pd.io.json.json_normalize(bib_dict,'identifier',['publisher', 'title', ['journal','name', 'name'], ['author'], 'year', 'number', 'volumne', ['link'], '_gddid', 'type', 'pages'], record_prefix='_', errors='ignore')

bibliography['link'] = bibliography['link'].astype(str)

url = bibliography['link'].str.split(", ", expand = True)

bibliography['link_url'] = url[0]
bibliography['link_type'] = url[1]

bibliography['link_url'] = bibliography['link_url'].replace("\[{'url': '", "", regex = True)\
                                                   .replace("'", "", regex = True)

bibliography['link_type'] = bibliography['link_type'].replace("'type': '", "", regex = True)\
                                                     .replace("'}]", "", regex = True)

bibliography['author'] = bibliography['author'].astype(str)
bibliography['author'] = bibliography['author'].replace("\[{'name': '", "", regex = True)\
                                               .replace("{'name': '", "", regex = True)\
                                               .replace("'},", ";", regex = True)\
                                               .replace("'}]", "", regex = True)
bibliography = bibliography[['_type', '_id', 'publisher', 'title', 'journal.name.name',	'author', 'year', 'number', 'volumne', '_gddid', 'type', 'pages', 'link_url', 'link_type']]

bibliography.head(3)

_type                            _id                    publisher  \
0   doi  10.1016/S0277-3791(99)00007-4                     Elsevier   
1   doi                10.1139/b92-002  Canadian Science Publishing   
2   doi   10.1016/0031-0182(80)90043-7                     Elsevier   

                                               title  \
0  Palaeoclimate, chronology and vegetation histo...   
1  Holocene history of forest trees on the Bruce ...   
2  Glacial sequence and environmental history in ...   

                                   journal.name.name  \
0                         Quaternary Science Reviews   
1                         Canadian Journal of Botany   
2  Palaeogeography, Palaeoclimatology, Palaeoecology   

                                              author  year number volumne  \
0  Huntley, B.; Watts, W.A.; Allen, J.R.M.; Zolit...  1999      7     NaN   
1                                     Bennett, K. D.  1992      1     NaN   
2  Van Der Hammen, T.; Barelds, J.; De Jong, H.; ...  1980    NaN     NaN   

                     _gddid     type     pages  \
0  550453fde1382326932d85f7  article  945--960   
1  578b5aabcf58f1587003efba  article     6--18   
2  54b43244e138239d868493cd  article  247--340   

                                            link_url  link_type  
0  http://www.sciencedirect.com/science/article/p...  publisher  
1  http://www.nrcresearchpress.com/doi/abs/10.113...  publisher  
2  http://www.sciencedirect.com/science/article/p...  publisher

In [9]:
bibliography.describe(include = 'all')

_type                               _id publisher  \
count    221                               221       221   
unique     1                               221         6   
top      doi  10.1111/j.1502-3885.2010.00159.x  Elsevier   
freq     221                                 1       138   

                                                    title  \
count                                                 221   
unique                                                221   
top     Charcoal signatures defined by multivariate an...   
freq                                                    1   

          journal.name.name          author  year number volumne  \
count                   221             221   221    183       0   
unique                   41             213    47     28       0   
top     Quaternary Research  Markgraf, Vera  2011      3     NaN   
freq                     44               3    14     40     NaN   

                          _gddid     type  pages  \
count                        221      221    216   
unique                       221        1    214   
top     550505d7e1382326932d8b07  article  1--13   
freq                           1      221      2   

                                                 link_url  link_type  
count                                                 221        220  
unique                                                221          1  
top     http://www.sciencedirect.com/science/article/p...  publisher  
freq                                                    1        220

## EDA

Reviewing our data includes skimming through some papers online and seeing if the data is consistent with our NLP Sentences dataframe.

From there, we can also visualize what we would like our model to predict: 'Location', 'Site Name', 'Age Span', and 'Site Description' from a "Human perspective".

### Finding Latitude and Longitud via REGEX 

After reviewing some PDF's and viewing the data parsing, I noticed some possible patterns for coordinates.

<img src="figures/img/00_snippet_of_pdf.png" alt="Drawing" style="width: 300px;"/>

Which translates to our pandas dataframe as:

![alt text](figures/img/01_snippet_in_sql.png "PandasVersion")

As a first problem, we can notice that the degrees symbol `°` was translated as a `3`. On some other occassions, it is confused as an `o` or a middle point `◦`. For `'`, this symbol got confused by the word `cents`.

I am still going to use normal REGEX to find coordinates in the most intuitive way. 

A coordinate for longitude is a number ranging from -180° to 180°. It has three components, `°, ',"` and it covers the position in E or W.

Latitude goes from -90° to 90°, has the same three components and goes from S to N.

Using REGEX, and assuming data consistency, the following code should be able to extract coordinates. 

In [10]:
dms_regex = r"([-]?\d{1,3}[°|′|\'|,]{0,3}\d{1,2}[,|′|\']{1,3}\d{0,2}[,|′|\']{1,4}[NESWnesw]?[\s|,|\']+?[-]?\d{1,3}[°|,|′|\']+\d{1,2}[,|′|\']+\d{0,2}[,|′|\'][,|′|\']{0,4}[NESWnesw]?)"

dd_regex = "[\\{,][-]?[1]?[0-9]{1,2}\\.[0-9]{1,}[,]?[NESWnesw],"

In [27]:
# TODO Find REGEX of couples of numbers (123, 234)
nlp_sentences['words_as_string'] = nlp_sentences['words'].apply(lambda x: ','.join(map(str, x)))
nlp_sentences['dms_regex'] = nlp_sentences['words_as_string'].str.findall(dms_regex)


#nlp_sentences[['dms_regex', 'words']]


#nlp_sentences['dms_regex'] = nlp_sentences['dms_regex'].astype(str)
#nlp_sentences['dms_regex'] = nlp_sentences['dms_regex'].replace(",", ";", regex = True)#\
                                                       #.replace(",", "", regex = True)

nlp_sentences[['words_as_string', 'words', 'dms_regex']].head(3)

words_as_string  \
0  Available,online,at,www.sciencedirect.com,Quat...   
1  The,Chihuahueños,Bog,record,extends,to,over,15...   
2  An,Artemisia,steppe,,,then,an,open,Picea,woodl...   

                                               words dms_regex  
0  [Available, online, at, www.sciencedirect.com,...        []  
1  [The, Chihuahueños, Bog, record, extends, to, ...        []  
2  [An, Artemisia, steppe, ,, then, an, open, Pic...        []

In [28]:
#nlp_sentences[['words_as_string', 'words', 'dms_regex']].describe()

The `dms_regex` only returned 95 unique values. That's to say that from the 111965 sentences, only 95 possess something similar to the regex we are looking for.  

Let's see an example of a regex straction as a sanity check that we are getting the values that we need. 

In [29]:
italy_example = nlp_sentences[nlp_sentences['_gddid'] == '550453fde1382326932d85f7']
italy_example = italy_example[italy_example['sentid'] == 10]
italy_example[['words', 'words_as_string', 'dms_regex']]

words  \
94875  [Introduction, Laghi, di, Monticchio, are, two...   

                                         words_as_string  \
94875  Introduction,Laghi,di,Monticchio,are,two,maar,...   

                         dms_regex  
94875  [403,56,40,N,,,153,36,48,E]

### Problems with the REGEX approach

As seen before, extracting information using REGEX is very complicated. We would require a lot of queries that would retrieve each particular case.

A possible alternative could be using NER, Named Entity Recognition.

In [30]:
def order_article(article_id):
    '''
    Function to find an article by its gddid in the NLP sentences and have it displayed in order
    
    Keyword arguments:
    article_id -- gddid
    
    Returns:
    article ordered by sentence index
    '''
    article = nlp_sentences[nlp_sentences['_gddid'] == article_id]
    return article[['sentid', 'words']].sort_values(by = 'sentid')

Article 01  
Skimmed info: 
 * `Article Name:` Palaeoclimate_chronology_and_vegetation_history_of_the_Weichselian_Lateglacial
 * `Link`: http://www.sciencedirect.com/science/article/pii/S0277379199000074  
 * `Site Name`:  Laghi di Monticchio
 * `Location`:  40° 56' 40" N, 15° 36' 48" E
 * `Age Span`: 
 * `Site Descriptions`: The record from Lago Grande di Monticchio now adds macrofossil evidence of the presence of relatively boreal tree taxa during the Lateglacial interstadial and provides a chronology based upon annually laminated sediments.

In [31]:
# Article 01 
order_article('550453fde1382326932d85f7')

sentid                                              words
94865       1  [Quaternary, Science, Reviews, 18, -LRB-, 1999...
94866       2  [Plant, macrofossil, data, from, one, core, pr...
94867       3  [Quantitative, palaeoclimate, reconstructions,...
94868       4  [The, chronological, framework, for, Lateglaci...
94870       5  [Surface, conditions, in, the, North, Atlantic...
...       ...                                                ...
95325     448  [Vegetation, history, and, climate, of, the, l...
95326     449  [Quaternary, Science, Reviews, ,, 15, ,, 133, ...
95327     450  [Zolitschka, ,, B., ,, &, Negendank, ,, J., F....
95329     451  [Sedimentology, ,, dating, and, palaeoclimatic...
95330     452  [Quaternary, Science, Reviews, ,, 15, ,, 101, ...

[452 rows x 2 columns]

Article 02  
Skimmed info:  
* `Article:` Development of the Mixed Connifer Forest in Northern New Mexico and its relationship to Holocene environmental change
* `Link`: http://www.sciencedirect.com/science/article/pii/S0033589407001512
* `Site Name`:  Chihuahuenos Bog 
* `Location`:  ??? No Given Coordinates 
* `Age Span`:   
* `Site Descriptions`:  Chihuahueños Bog is located within the mixed conifer forest and has affinities to the flora of the southern Rocky Mountains(Foxx and Tierney, 1985; Allen, 2004). Although the area immediately surrounding the bog has been heavily logged in recent years, nearly every conifer species growing in montaneforests of the Jemez mountains is represented either along theshore of the bog or in the surrounding uplands.  

In [32]:
# Article 2
order_article('54b43266e138239d8684efed')

sentid                                              words
0         1  [Available, online, at, www.sciencedirect.com,...
1         2  [The, Chihuahueños, Bog, record, extends, to, ...
2         3  [An, Artemisia, steppe, ,, then, an, open, Pic...
3         4  [C/N, ratios, ,, δ13C, and, δ15N, values, indi...
4         5  [Higher, percentages, of, aquatic, algae, and,...
..      ...                                                ...
605     572  [Is, the, Valles, caldera, entering, a, new, c...
144     573                  [Geology, 23, ,, 411, --, 414, .]
607     574  [Wright, Jr., ,, H.E., ,, Bent, ,, A.M., ,, Ha...
608     575  [Present, and, past, vegetation, of, the, Chus...
609     576  [Geological, Society, of, America, Bulletin, 8...

[576 rows x 2 columns]

Article 03  
Skimmed info:  
* `Link`: http://dx.doi.org/10.1130/g35541.1  # No Full access to article
* `Site Name`:  Cahokia region, Illinois, USA
* `Location`:  N38°42'00" - N38°42'00", W90°04'60" - W90°04'60" - Not found on text (yet).
* `Age Span`:   
* `Site Descriptions`: Record from Horseshoe Lake, an oxbow lake in the central Mississippi River valley that is adjacent to the Cahokia site (Illinois, USA), the largest prehistoric settlement north of Mexico.   

In [33]:
# Article 3
order_article('57c5b941cf58f1338eaddb5b')

sentid                                              words
67708       1  [A, record, of, sustained, prehistoric, and, h...
67709       2  [Here, we, report, a, high-resolution, and, mu...
67710       3  [Palynological, and, carbon, isotope, data, do...
67712       4  [Rapid, forest, clearance, was, followed, clos...
67713       5  [Agricultural, intensification, that, included,...
...       ...                                                ...
67845     123  [Simon, ,, M.L., ,, and, Parker, ,, K.E., ,, 2...
67846     124  [Smith, ,, B.D., ,, and, Yarnell, ,, R.A., ,, ...
67848     125  [Sugita, ,, S., ,, 1993, ,, A, model, of, poll...
67849     126  [Trubitt, ,, M.B.D., ,, 2000, ,, Mound, buildi...
67850     127  [Woods, ,, W.I., ,, 2004, ,, Population, nucle...

[127 rows x 2 columns]

Article 04  
Skimmed info:  
* `Link:`http://www.tandfonline.com/doi/abs/10.1080/00173130902965157
* `Site Name`:  Begbunar
* `Location`:  42°09′ N, 22° 33′ E; 1750 ; in SQL file shows as: `42,°,09,cents,N,,,22,°,33,cents,E`
* `Age Span`:   
* `Site Descriptions`: Located in the central treeless zone of the Osogovo mountains, which are situated at the border between south-western Bulgaria and the north-eastern former Yugoslav Republic of Macedonia.

In [34]:
# Article 4
order_article('58d29193cf58f14928755ba5')

sentid                                              words
110659       1  [Grana, ISSN, :, 0017-3134, -LRB-, Print, -RRB...
110660       2  [Peat-bog, Begbunar, -LRB-, Osogovo, Mountains...
110661       3  [Peat-bog, Begbunar, -LRB-, Osogovo, Mountains...
110662       4  [Submit, your, article, to, this, journal, Art...
110665       5  [Peat-bog, Begbunar, -LRB-, Osogovo, Mountains...
...        ...                                                ...
110740      80                                           [Ann, .]
110741      81                                  [Sofia, Univ., .]
110742      82                                           [Fac, .]
110743      83                                          [Geol, .]
110744      84  [Geogr., ,, 85, ,, 181, --, 198, -LRB-, in, Bu...

[84 rows x 2 columns]

In [35]:
# Try to see the exact location. Wont show last few words.
pd.DataFrame(order_article('58d29193cf58f14928755ba5').iloc[4])

110665
sentid                                                  5
words   [Peat-bog, Begbunar, -LRB-, Osogovo, Mountains...

Article 05  
Skimmed info:  
* `Link`: http://dx.doi.org/10.1130/0016-7606(1997)109<1306:tfalqp>2.3.co
* `Site Name`:  Cottonwood III, Cottonwood IV, Red Lady, Red Well, Splains
* `Location`:  38°49'50' 106°24'45', 38°49'50' 106°24'45', 38°52'50' 107°2'30', 38°52'50' 107°2'30',38°53'40' 107°3'15', 38°52'28' 107°05', 38°50' 107°4'30', 38°50' 107°4'30', 38°52' 107°2'30', 38°45' 106°50', 38°45' 106°50'
* `Age Span`:   17000 - 400
* `Site Descriptions`:   The lower timberline or lower forest border is the lower elevational limit of a continuous forest belt. In the Rocky Mountains it forms a sharp contrast with shrublands or grasslands in valley bottoms. Trees may extend below the lower timberline in favorable habitats, along streams or on rocky outcrops.  

In [36]:
# Article 5
order_article('57928e07cf58f133d1c26609')

sentid                                              words
39606       1  [Timberline, fluctuations, and, late, Quaterna...
39607       2  [By, tracking, climatically, sensitive, forest...
39608       3  [Pollen, data, suggest, that, prior, to, 11, 0...
39609       4  [The, inferred, climate, was, 2Ð5, ¡, C, coole...
39611       5  [Abies, -LRB-, fir, -RRB-, increased, in, abun...
...       ...                                                ...
40107     448  [Weber, ,, W., A., ,, 1987, ,, Colorado, flora...
40108     449  [Whitlock, ,, C., ,, 1993, ,, Postglacial, veg...
40110     450  [Whitlock, ,, C., ,, and, Bartlein, ,, P., J.,...
40111     451  [Wright, ,, H., E., ,, Jr., ,, 1983, ,, Late-Q...
40112     452  [MANUSCRIPT, RECEIVED, BY, THE, SOCIETY, MAY, ...

[452 rows x 2 columns]

In [37]:
# Print the exact line where locations are.
pd.DataFrame(order_article('57928e07cf58f133d1c26609').iloc[179])

39801
sentid                                                180
words   [LOCATION, AND, AGE, OF, FOSSIL, STUDY, SITES,...

In [38]:
regex_example = nlp_sentences[nlp_sentences['_gddid'] == '57928e07cf58f133d1c26609']
regex_example = regex_example[regex_example['sentid'] == 180]
regex_example[['words', 'words_as_string', 'dms_regex']]

words  \
39801  [LOCATION, AND, AGE, OF, FOSSIL, STUDY, SITES,...   

                                         words_as_string  \
39801  LOCATION,AND,AGE,OF,FOSSIL,STUDY,SITES,Core,Lo...   

                                               dms_regex  
39801  [38,°,49,′,50,′,′,106,°,24,′,45,′,′,, 38,°,49,...

In [39]:
# Find the article in the bibliography by gddid to retrieve Article Name

bibliography[bibliography['_gddid'] == '57928e07cf58f133d1c26609']

_type                                              _id publisher  \
93   doi  10.1130/0016-7606(1997)109<1306:tfalqp>2.3.co;2       GSA   

                                                title  \
93  Timberline fluctuations and late Quaternary pa...   

                         journal.name.name             author  year number  \
93  Geological Society of America Bulletin  Fall, Patricia L.  1997     10   

   volumne                    _gddid     type       pages  \
93     NaN  57928e07cf58f133d1c26609  article  1306--1320   

                                             link_url  link_type  
93  http://dx.doi.org/10.1130/0016-7606(1997)109<1...  publisher

## Linking Neotoma DB to extract locations

In [288]:
neotoma = pd.read_csv("../Do_not_commit_data/data-1590729612420.csv")
neotoma = neotoma[['siteid', 'sitename', 'longitudeeast', 'latitudenorth', 'longitudewest', 'latitudesouth', 'sitedescription', 'doi']]

neotoma.head(3)

siteid       sitename  longitudeeast  latitudenorth  longitudewest  \
0  10330   Lac du Sommet -70.66468       47.71662      -70.66573        
1  10330   Lac du Sommet -70.66468       47.71662      -70.66573        
2  1729    Myrtle Lake   -93.37853       47.98645      -93.39207        

   latitudesouth  \
0  47.71382        
1  47.71382        
2  47.97876        

                                                                                                                                                                                                                                                                                                             sitedescription  \
0  The small shallow Lac du Sommet (0.02 km2, 4 m maximum depth, elevation of 830 m a.s.l., 47°43′N, 70°40′W) is located in the boreal forest north of the St Lawrence Estuary (Figure 1). It is of glacial origin and situated on granitic-gneissic bedrock of the Canadian Precambrian Shield in the Laurentian Mountains.   
1  The small shallow Lac du Sommet (0.02 km2, 4 m maximum depth, elevation of 830 m a.s.l., 47°43′N, 70°40′W) is located in the boreal forest north of the St Lawrence Estuary (Figure 1). It is of glacial origin and situated on granitic-gneissic bedrock of the Canadian Precambrian Shield in the Laurentian Mountains.   
2  Lake surrounded by peatland. Physiography: Red Lake lowlands. Surrounding vegetation: Sphagnum, Piceto-Chamaedophnetum.                                                                                                                                                                                                     

                        doi  
0  10.1177/0959683611400199  
1  10.1177/0959683611400199  
2  10.1139/b68-190

In [289]:
nlp_bib = nlp_sentences.merge(bibliography, on = '_gddid')
nlp_bib = nlp_bib.rename(columns={'_id': 'doi'})
nlp_bib.head(3)

_gddid  sentid  \
0  54b43266e138239d8684efed  1        
1  54b43266e138239d8684efed  2        
2  54b43266e138239d8684efed  3        

                                                                                                                                                                                                                                                                                                                                                                                                         wordidx  \
0  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, ...]   
1  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]                                                                                                                                                                                                                                                                                                                                                                         
2  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26]                                                                                                                                                                                                                                                                                                                 

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       words  \
0  [Available, online, at, www.sciencedirect.com, Quaternary, Research, 69, -LRB-, 2008, -RRB-, 263, --, 275, www.elsevier.com/locate/yqres, Development, of, the, mixed, conifer, forest, in, northern, New, Mexico, and, its, relationship, to, Holocene, environmental, change, R., Scott, Anderson, a, ,, b, ,, ,, Renata, B., Jass, b, ,1, ,, Jaime, L., Toney, b, ,2, ,, Craig, D., Allen, c, ,, Luz, M., Cisneros-Dozal, d, ,, Marcey, Hess, d, ,, Jeff, Heikoop, d, ,, Julianna, Fessenden, d, a, Center, for, Environmental, Sciences, &, Education, ,, Box, 5694, ,, Northern, Arizona, University, ,, Flagstaff, ,, AZ, 86011, ,, USA, b, Quaternary, Sciences, Program, &, Bilby, Research, ...]   
1  [The, Chihuahueños, Bog, record, extends, to, over, 15,000, cal, yr, BP, .]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 
2  [An, Artemisia, steppe, ,, then, an, open, Picea, woodland, grew, around, a, small, pond, until, ca., 11,700, cal, yr, BP, when, Pinus, ponderosa, became, established, .]          

In [290]:
print(nlp_bib.shape)

(111965, 24)


In [291]:
nlp_bib[['words_as_string', 'journal.name.name', 'dms_regex']].head(3)

words_as_string  \
0  Available,online,at,www.sciencedirect.com,Quaternary,Research,69,-LRB-,2008,-RRB-,263,--,275,www.elsevier.com/locate/yqres,Development,of,the,mixed,conifer,forest,in,northern,New,Mexico,and,its,relationship,to,Holocene,environmental,change,R.,Scott,Anderson,a,,,b,,,,,Renata,B.,Jass,b,,1,,,Jaime,L.,Toney,b,,2,,,Craig,D.,Allen,c,,,Luz,M.,Cisneros-Dozal,d,,,Marcey,Hess,d,,,Jeff,Heikoop,d,,,Julianna,Fessenden,d,a,Center,for,Environmental,Sciences,&,Education,,,Box,5694,,,Northern,Arizona,University,,,Flagstaff,,,AZ,86011,,,USA,b,Quaternary,Sciences,Program,&,Bilby,Research,Center,,,Box,6013,,,Northern,Arizona,University,,,Flagstaff,,,AZ,86011,,,USA,c,U.S.,Geological,Survey,,,Jemez,Mountains,Field,Station,,,Bandelier,National,Monument,,,HCR-1,,,Box,1,,,#,15,,,Los,Alamos,,,NM,87544,,,USA,d,Earth,&,Environmental,Sciences,Division,,,Hydrology,,,Geochemistry,&,Geology,Group,,,EES-6,,,MS-D462,,,Los,Alamos,National,Laboratory,,,Los,Alamos,,,NM,87545,,,USA,Received,28,April,2007,Available,online,29,January,2008,Abstract,Chihuahueños,Bog,-LRB-,2925,m,-RRB-,in,the,Jemez,Mountains,of,northern,New,Mexico,contains,one,of,the,few,records,of,late-glacial,and,postglacial,development,of,the,mixed,conifer,forest,in,southwestern,North,America,.   
1  The,Chihuahueños,Bog,record,extends,to,over,15,000,cal,yr,BP,.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              
2  An,Artemisia,steppe,,,then,an,open,Picea,woodland,grew,around,a,small,pond,until,ca.,11,700,cal,yr,BP,when,Pinus,ponderosa,became,established,.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             

     journal.name.name dms_regex  
0  Quaternary Research  []        
1  Quaternary Research  []        
2  Quaternary Research  []

Let's now try to extract Sitenames using Neotoma database. For this, I will group by doi and then, list all sitenames that are mentioned in each doi.

In [292]:
by_doi = neotoma.groupby('doi')
neotoma_sitenames = by_doi.apply(lambda x: list(x['sitename']))
neotoma_siteid = by_doi.apply(lambda x: list(x['siteid']))
neotoma_longeast = by_doi.apply(lambda x: list(x['longitudeeast']))
neotoma_latnorth = by_doi.apply(lambda x: list(x['latitudenorth']))

neotoma_joined_df = neotoma_sitenames.to_frame('sitenames').join(neotoma_siteid.to_frame('siteid')).join(neotoma_longeast.to_frame('longeast')).join(neotoma_latnorth.to_frame('latnorth'))
neotoma_joined_df = neotoma_joined_df.reset_index()

#neotoma_joined_df['intlongeast'] = [int(i) for neotoma_joined_df['longeast'][i]]

#print(neotoma_joined_df['longeast'][1].astype(int))
#[round(x) for x in myList]
neotoma_joined_df.head(5)


doi  \
0  0.1139/e80-122                                                      
1  10.1002/(SICI)1099-1417(199607/08)11:4<327::AID-JQS257>3.0.CO;2-G   
2  10.1002/(SICI)1099-1417(199701/02)12:1<15::AID-JQS283>3.0.CO;2-T    
3  10.1002/(SICI)1099-1417(199801/02)13:1<43::AID-JQS348>3.0.CO;2-G    
4  10.1002/1099-1417(200010)15:7<687::AID-JQS551>3.0.CO;2-6            

                                     sitenames              siteid  \
0  [East Baltic Bog, Mermaid Bog, Portage Bog]  [763, 1684, 1903]    
1  [Lundin Tower]                               [3256]               
2  [Spruce Pond, Spruce Pond, Spruce Pond]      [2508, 2508, 2508]   
3  [Pantano de Genagra, Pantano de Genagra]     [14030, 14030]       
4  [Laguna Loma Linda, Laguna Loma Linda]       [13971, 13971]       

                            longeast  \
0  [-62.15, -63.025, -64.075]          
1  [-2.97206]                          
2  [-74.18236, -74.18236, -74.18236]   
3  [-76.60428, -76.60428]              
4  [-73.35661, -73.35661]              

                                                       latnorth  
0  [46.408333, 46.25, 46.673611]                                 
1  [56.21529]                                                    
2  [41.237770000000005, 41.237770000000005, 41.237770000000005]  
3  [2.4751, 2.4751]                                              
4  [3.30805, 3.30805]

From Neotoma DB, we have obtained a dataframe grouped by DOI that lists each `sitename`, `siteid`, `longitude east`, and `latitunde north`. Now, let's join this information to the NLP_Bib dataframe. For now, let's do to the one that coordinates have been obtained with REGEX just to see if some of the numbers and sitenames make sense. 

In [293]:
nlp_bib_neotoma = nlp_bib.merge(neotoma_joined_df, on = 'doi')
print(nlp_bib_neotoma.shape)

(106640, 28)


In [294]:
# Filter df to just frames that has REGEX longer than one
mask = (nlp_bib_neotoma['dms_regex'].str.len() > 0)
nlp_bib_neotoma_regex = nlp_bib_neotoma.loc[mask]
nlp_bib_neotoma_regex[['words','dms_regex']].head(100)

print(nlp_bib_neotoma_regex.shape)

(93, 28)


In [295]:
nlp_bib_neotoma_regex = nlp_bib_neotoma_regex[['_gddid', 'sentid', 'doi', 'words', 'words_as_string', 'dms_regex', 'sitenames', 'latnorth', 'longeast']]
nlp_bib_neotoma_regex = nlp_bib_neotoma_regex.reset_index()
nlp_bib_neotoma_regex.head(1)

index                    _gddid  sentid              doi  \
0  3314   578a1f6ecf58f1b54b810cd9  94      10.1139/b85-070   

                                                                                                                                                                                                                                               words  \
0  [Lake, Hope, Simpson, -LRB-, 52, ', 27, ', N, ,, 56, ', 26, ', W, ,, elevation, 295, m, -RRB-, lies, 20, km, southwest, of, Port, Hope, Simpson, in, a, steep-sided, eastwest-trending, valley, that, is, drained, by, the, St., Lewis, River, .]   

                                                                                                                                                                                          words_as_string  \
0  Lake,Hope,Simpson,-LRB-,52,',27,',N,,,56,',26,',W,,,elevation,295,m,-RRB-,lies,20,km,southwest,of,Port,Hope,Simpson,in,a,steep-sided,eastwest-trending,valley,that,is,drained,by,the,St.,Lewis,River,.   

                     dms_regex                               sitenames  \
0  [52,',27,',N,,,56,',26,',W]  [Lake Hope Simpson, 18 [Moraine Lake]]   

             latnorth              longeast  
0  [52.45, 52.266667]  [-56.433333, -58.05]

In [ ]:
any( k in s for k in keywords )

We can see that most of the REGEX indeed pulled out the same locations for latitude and longitudes as in the Neotoma database.

In [296]:
nlp_bib_neotoma[['_gddid', 'sentid', 'doi', 'words', 'dms_regex', 'sitenames', 'latnorth', 'longeast']].head(2)

_gddid  sentid                          doi  \
0  54b43266e138239d8684efed  1       10.1016/j.yqres.2007.12.002   
1  54b43266e138239d8684efed  2       10.1016/j.yqres.2007.12.002   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       words  \
0  [Available, online, at, www.sciencedirect.com, Quaternary, Research, 69, -LRB-, 2008, -RRB-, 263, --, 275, www.elsevier.com/locate/yqres, Development, of, the, mixed, conifer, forest, in, northern, New, Mexico, and, its, relationship, to, Holocene, environmental, change, R., Scott, Anderson, a, ,, b, ,, ,, Renata, B., Jass, b, ,1, ,, Jaime, L., Toney, b, ,2, ,, Craig, D., Allen, c, ,, Luz, M., Cisneros-Dozal, d, ,, Marcey, Hess, d, ,, Jeff, Heikoop, d, ,, Julianna, Fessenden, d, a, Center, for, Environmental, Sciences, &, Education, ,, Box, 5694, ,, Northern, Arizona, University, ,, Flagstaff, ,, AZ, 86011, ,, USA, b, Quaternary, Sciences, Program, &, Bilby, Research, ...]   
1  [The, Chihuahueños, Bog, record, extends, to, over, 15,000, cal, yr, BP, .]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 

  dms_regex                             sitenames            latnorth  \
0  []        [Chihuahueños Bog, Chihuahueños Bog]  [36.0491, 36.0491]   
1  []        [Chihuahueños Bog, Chihuahueños Bog]  [36.0491, 36.0491]   

                   longeast  
0  [-106.50963, -106.50963]  
1  [-106.50963, -106.50963]

In [320]:
#nlp_bib_neotoma['words_l'] = nlp_bib_neotoma['words'].apply(lambda col: col.str.lower())

nlp_bib_neotoma['words_l']  = [s.lower() for s in nlp_bib_neotoma['words']]
nlp_bib_neotoma[['words', 'words_l']].head(2)



AttributeError: 'list' object has no attribute 'lower'

In [303]:
any(item in nlp_bib_neotoma_regex['words'][48] for item in nlp_bib_neotoma_regex['sitenames'][48])

False

In [321]:
any( k in nlp_bib_neotoma_regex['words'] for k in nlp_bib_neotoma_regex['sitenames'])

TypeError: unhashable type: 'list'

In [304]:
nlp_bib_neotoma_regex['sitenames'][48]

['Uddelermeer', 'Bleekemeer', 'Mekelermeer']

In [305]:
for i in nlp_bib_neotoma_regex['words'][48]:
    print(i)

The
presence
of
a
-LRB-
or
more
-RRB-
hiatus
-LRB-
-
es
-RRB-
in
the
sequenceis
responsible
for
the
fact
that
no
accurate
estimation
could
be
made
BOREAS
17
-LRB-
1988
-RRB-
Biostra
-
t
igraphy
UDDELERMEER
Late-Glacial
lake-level
fluctuations
423
BLEEKEMEER
MEKELERMEEfi
II
I
;
,
;
'
6
'
1
0
'
2
0
'
3
O
'
4
6
'
5
3
0
5
5
6
'
'
'
,6
'
;
o'240
ii
5
45
6
7
wlume
V.
orgonic
m
a
t
t
e
d
x
1
0
-
~
Fig.
10
.


In [270]:
print(type(nlp_bib_neotoma['sitenames'][48]))
print(type(nlp_bib_neotoma['words'][48]))
print(type(nlp_bib_neotoma['words_as_string'][48]))

<class 'str'>
<class 'str'>
<class 'str'>


In [266]:
#for i in range(len(nlp_bib_neotoma.index)):
#    nlp_bib_neotoma['sitename_in_wordsstr'] = any(item in nlp_bib_neotoma['words_as_string'][i] for item in nlp_bib_neotoma['sitenames'][i])
    
    
    

In [265]:
#nlp_bib_neotoma.head(299)